docs/protocol/extensions/struct-dtypes/v1.0.rst

```diff
@@ -0,0 +1,127 @@
+=======================
+
+This document is a **Work in Progress**. It may be updated, replaced
+or obsoleted by other documents at any time. It is inappapropriate to
+cite this document as other than work in progress.
+
+Comments, questions or contributions to this document are very
+welcome. Comments and questions should be raised via `GitHub issues
+<https://github.com/zarr-developers/zarr-specs/labels/struct-dtypes-v1.0>`_. When
+raising an issue, please add the label "struct-dtypes-v1.0".
+
+This document was produced by the `Zarr core development team
+<https://github.com/orgs/zarr-developers/teams/core-devs>`_.
+
+
+Document conventions
+====================
+
+Conformance requirements are expressed with a combination of
+descriptive assertions and [RFC2119]_ terminology. The key words
+"MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in the normative
+parts of this document are to be interpreted as described in
+[RFC2119]_. However, for readability, these words do not appear in all
+uppercase letters in this specification.
+
+All of the text of this specification is normative except sections
+explicitly marked as non-normative, examples, and notes. Examples in
+this specification are introduced with the words "for example".
+
+
+Extension data types
+====================
+
+NumPy allows representation of `Structured Arrays`_ where each element of the
+array is actually some combination of fields, each of which may have its own
+unique data type. NumPy's Record Arrays (``numpy.recarray``) also use this data type. The actual data is stored as an opaque seqeuence of bytes
+(i.e. a structure) as represented by (``numpy.void``) and thus the string
+representation of this dtype in NumPy is ``'|Vn'`` where ``n`` is some integer
+number of bytes. In order to be able to properly interpret data of this type
+if is necessary to store information on the fields
+
+A concrete example of such an array from the NumPy docs is::
+
+    dogs = np.array([('Rex', 9, 81.0), ('Fido', 3, 27.0)],
+                    dtype=[('name', 'U10'), ('age', 'i4'), ('weight', 'f4')])
+
+where here ``dogs.dtype.kind`` is 'V' and ``dogs.dtype.str`` is ``'|V48'``
+indicating the 48 bytes are needed to store each element (4 bytes each for
+``age`` and ``weight`` and 4 * 10 = 40 bytes for a 10-character UTF-32
+``name``). If we were to read such a sequence of bytes from a Zarr array, we
+need the dtype description to know how to properly interpret this sequence of
+48 bytes. The NumPy dtype object has a ``descr`` attribute that describes this.
+In this case ``dogs.dtype.descr`` is ``[('name', '<U10'), ('age', '<i4'), ('weight', '<f4')]``.
+
+
+Data Types added by this extension
+==================================
+
+.. list-table:: Data types
+   :header-rows: 1
+
+   * - Identifier
+     - Numerical type
+     - Size (no. bytes)
+     - Byte order
+   * - list of (<name>, <type>) tuples
```

Comment:
NumPy (and zarr v2) supports a third variant, where you have `(name, type, shape)`.  The shape specifies the dimensions of an "inner" array of elements.

Reply:
Yeah, I had forgotten to list that one and will update it here to describe an optional [, shape]. I think we do test it already.